### PR TITLE
New elements in GeoMechanicsApplication

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**📝 Description**
In this PR the following elements are added:

- A 2-dimensional 3-noded beam element
- steady-state interface elements (pipe)
- transient interface elements (pipe)

In addition some minor modifications have been done and more tests are added.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_Pw_scheme.hpp [hunk 1]
```diff
@@ -135,8 +135,7 @@ class NewmarkQuasistaticPwScheme : public NewmarkQuasistaticUPwScheme<TSparseSpa
         ModelPart::NodesContainerType::iterator node_begin = rModelPart.NodesBegin();
 
         #pragma omp parallel for private(DeltaPressure)
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_Pw_scheme.hpp:137]: parallel utils :)

File: applications/GeoMechanicsApplication/custom_processes/apply_phreatic_line_pressure_table_process.hpp [hunk 2]
```diff
@@ -79,14 +75,10 @@ class ApplyPhreaticLinePressureTableProcess : public ApplyConstantPhreaticLinePr
 
         const double Time = mrModelPart.GetProcessInfo()[TIME]/mTimeUnitConverter;
         array_1d<double, 2> deltaH;
-        for (unsigned int i=0; i < mpTable.size(); ++i)
-        {
-            if (mpTable[i] == NULL)
-            {
+        for (unsigned int i=0; i < mpTable.size(); ++i) {
+            if (mpTable[i] == NULL) {
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_phreatic_line_pressure_table_process.hpp:79]:
```suggestion
            if (!mpTable[i]) {
```

File: applications/GeoMechanicsApplication/custom_processes/apply_phreatic_line_pressure_table_process.hpp [hunk 3]
```diff
@@ -40,15 +40,11 @@ class ApplyPhreaticLinePressureTableProcess : public ApplyConstantPhreaticLinePr
...
             unsigned int TableId = rParameters["table"][i].GetInt();
-            if (TableId > 0)
-            {
+            if (TableId > 0) {
                 mpTable[i] = model_part.pGetTable(TableId);
-            }
-            else
-            {
+            } else {
                 mpTable[i] = NULL;
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_phreatic_line_pressure_table_process.hpp:48]:
```suggestion
                mpTable[i] = nullptr;
```

more C++11

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 4]
```diff
@@ -2510,25 +2510,17 @@ void SmallStrainUPwDiffOrderElement::
...
-
-    KRATOS_ERROR_IF(detJ < 0.0)
-     << "ERROR:: ELEMENT ID: "
-     << this->Id()
-     << " INVERTED. DETJ: "
-     << detJ
-     << std::endl;
+    if(detJ < 0.0) {
+        std::cerr << "Error: Element " << this->Id() << " is inverted. DetJ: " << detJ << std::endl 
+        << "This usually indicates the deformations are too large for the mesh size." << std::endl;
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:2515]: I would remove this and add it only to the error message

File: applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.hpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+#if !defined(KRATOS_GEO_TRANSIENT_PW_INTERFACE_ELEMENT_H_INCLUDED )
+#define  KRATOS_GEO_TRANSIENT_PW_INTERFACE_ELEMENT_H_INCLUDED
+
+// Project includes
+#include "includes/serializer.h"
+
+// Application includes
+#include "custom_elements/U_Pw_small_strain_interface_element.hpp"
+#include "custom_utilities/interface_element_utilities.hpp"
+#include "geo_mechanics_application_variables.h"
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.hpp:22]: I think those can be moved to the cpp

File: applications/GeoMechanicsApplication/tests/test_curved_beam_elements.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import sys
+import os
+
+sys.path.append(os.path.join('..', '..', '..'))
+sys.path.append(os.path.join('..', 'python_scripts'))
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/tests/test_curved_beam_elements.py:5]: I STRONGLY recommend against doing this
Modifying the system path remains after running the test and might lead to some hard-to-detect import problems!

Why do you need this?
You should be able to import all the python scripts in the usual way

File: applications/GeoMechanicsApplication/tests/test_curved_beam_elements.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        self.assertAlmostEqual(expected_value, x_displacements[node_number-1], 6)
+
+
+if __name__ == '__main__':
+    suites = KratosUnittest.KratosSuites
+    smallSuite = suites['small'] # These tests are executed by the continuous integration tool
+    smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([KratosGeoMechanicsCurvedBeamElementTests]))
+    allSuite = suites['all']
+    allSuite.addTests(smallSuite)
+    KratosUnittest.runTests(suites)
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/tests/test_curved_beam_elements.py:42]:
```suggestion
    KratosUnittest.main()
```

this should do the same, check the tests in the structural

File: applications/GeoMechanicsApplication/tests/test_transient_groundwater_flow.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import sys
+import os
+
+sys.path.append(os.path.join('..', '..', '..'))
+sys.path.append(os.path.join('..', 'python_scripts'))
+
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/tests/test_transient_groundwater_flow.py:6]:
```suggestion
```

File: applications/GeoMechanicsApplication/tests/test_transient_groundwater_flow.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        water_pressure = test_helper.get_water_pressure(stages[-1])
+
+        p_height_1_72_m = water_pressure[87]
+
+        self.assertAlmostEqual(2.6423560731957005, p_height_1_72_m)
+
+
+if __name__ == '__main__':
+    suites = KratosUnittest.KratosSuites
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/tests/test_transient_groundwater_flow.py:47]: same as above

**Review by philbucher — APPROVED** [review]
Some more comments:
- I recommend using the `ParallelUtilities` where possible, esp since they support exceptions now :)
- Can you reduce the mdpa sizes or reuse the existing ones? The GeomechanicsApp is already the 5th largest one with 13MB :/ this is ~5% of the entire Kratos source code

**Comment by Vahid-Galavi** [comment]
Thanks @philbucher for the review and nice suggestions. I will consider them in the next iteration.

**Pull request merged**